File: samples/shorturls/app/controllers/Application.scala
package controllers

import play.api._
import play.api.mvc._
import models.ShortURLs
import play.api.libs.json.Json
import play.api.Play.current
import models.ShortURLs._

object Application extends Controller {
  
  def index = Action {
    Ok(views.html.index())
  }

  def all = Action {
    Ok(views.html.all())
  }

  def goTo(id: String) = Action {
    Async {
      ShortURLs.findById(id).map { maybe =>
        maybe.map( url => Redirect(url.originalUrl) ).getOrElse(NotFound)
      }
    }
  }
}

File: samples/java/shorturls/app/controllers/Application.java
package controllers;

import play.libs.F;
import play.mvc.*;

import models.*;

public class Application extends Controller {

    public static Result index() {
        return ok(views.html.index.render());
    }

    public static Result goTo(String id) {
        return async(
            ShortURL.findById(id).map(new F.Function<ShortURL, Result>() {
                @Override
                public Result apply(ShortURL shortURL) throws Throwable {
                    if (shortURL == null) {
                        return notFound();
                    }
                    return redirect(shortURL.originalUrl);
                }
            })
        );
    }
}